[WIP][python3] Upgrade to 3.8.6

smaarn opened the pull request:
_Motivation:_  Upgrade python version to 3.8 (since it's now the oldest "stable" version)
_Linked issues:_  #4284 

### Checklist
- [x] Build rule `all-supported` completed successfully
- [ ] beets: Package upgrade completed successfully
- [ ] beets: New installation of package completed successfully
- [ ] borgbackup: Package upgrade completed successfully
- [ ] borgbackup: New installation of package completed successfully
- [ ] domoticz: Package upgrade completed successfully
- [ ] domoticz: New installation of package completed successfully
- [ ] ffsync: Package upgrade completed successfully
- [ ] ffsync: New installation of package completed successfully
- [ ] flexget: Package upgrade completed successfully
- [ ] flexget: New installation of package completed successfully
- [ ] homeassistant: Package upgrade completed successfully
- [ ] homeassistant: New installation of package completed successfully
- [ ] python: Package upgrade completed successfully
- [ ] python: New installation of package completed successfully
- [ ] python3: Package upgrade completed successfully
- [ ] python3: New installation of package completed successfully
- [ ] rdiff-backup: Package upgrade completed successfully
- [ ] rdiff-backup: New installation of package completed successfully
- [ ] sabnzbd: Package upgrade completed successfully
- [ ] sabnzbd: New installation of package completed successfully
- [ ] salt-minion: Package upgrade completed successfully
- [ ] salt-minion: New installation of package completed successfully
- [ ] vim: Package upgrade completed successfully
- [ ] vim: New installation of package completed successfully
- [ ] znc: Package upgrade completed successfully
- [ ] znc: New installation of package completed successfully

### Followups
- [x] Fix domoticz (see https://github.com/domoticz/domoticz/issues/3703)
- [x] Fix ffsync build
- [x] Fix mosquitto build
- [x] Fix homeassistant build
- [x] Fix znc build on `hi3535`
- [x] Fix rdiff-backup build on `evansport`
- [x] Fix homeassistant build on `x64`

hgy59 commented:
@smaarn can you please rebase to head@master to include the update of  cross/libffi to version 3.3?

hgy59 commented:
@smaarn as a learned on the home assistant update with #4149 there is an issue with cffi 1.14.4, so older compilers fail to build it.
The original bug has been fixed already in cffi:
https://foss.heptapod.net/pypy/cffi/-/commit/4855f749ad397f79867d214b2b516664bec14019)
But this fix is not yet released, so we still use cffi 1.14.1.
We should use cffi 1.14.1 here too, until a fixed version is released (will fix one kind of homeassistant build failures).

affected files
- native/python3/Makefile
- cross/cffi/Makefile

smaarn commented on mk/spksrc.wheel.mk:
```diff
@@ -63,9 +63,9 @@ build_wheel_target: $(PRE_WHEEL_TARGET)
 		if [ ! -z "$(CROSS_COMPILE_WHEELS)" ] ; then \
 			$(MSG) "Force cross-compile" ; \
 			if [ -z "$(CROSSENV)" ]; then \
-				$(RUN) _PYTHON_HOST_PLATFORM="$(TC_TARGET)" CFLAGS="$(CFLAGS) -I$(STAGING_INSTALL_PREFIX)/$(PYTHON_INC_DIR) $(WHEELS_CFLAGS)" LDFLAGS="$(LDFLAGS) $(WHEELS_LDFLAGS)" $(PIP_WHEEL) ; \
```
It turnsout that using `-I` wasn't appreciated in some cases (more specifically `gevent`) where it was failing, indicating that it should go through `CPPFLAGS` instead.

smaarn commented on spk/ffsync/Makefile:
```diff
@@ -40,6 +40,7 @@ include ../../mk/spksrc.spk.mk
 ffsync_extra_install:
 	install -m 755 -d $(STAGING_DIR)/var
 	install -m 644 src/ffsync.ini $(STAGING_DIR)/var/
+	install -m 644 src/requirements.txt $(STAGING_DIR)/share/wheelhouse/requirements.txt
```
Otherwise the `sed` operation below was failing. To be honest it's either we remove that line **and** the below line (the `sed` one). Or we keep that line or at least I need to know where that `requirements.txt` file would be supposed to be coming from.

smaarn commented on spk/ffsync/src/requirements.txt:
```diff
@@ -92,11 +92,11 @@ repoze.lru==0.6
 #requests==2.4.3
 simplejson==3.6.5
 six==1.8
-testfixtures==4.1.0
```
This was failing building. But since it's related to testing I figured we could dismiss it, right ?

smaarn commented on spk/ffsync/src/requirements.txt:
```diff
@@ -92,11 +92,11 @@ repoze.lru==0.6
 tokenlib==0.3.1
 translationstring==1.1
 umemcache==1.6.3
-unittest2==0.6
```
This was failing building. But since it's related to testing I figured we could dismiss it, right ?

smaarn commented:
Corresponding binaries have been published here: https://github.com/smaarn/spksrc/releases/tag/python-3.8.6-RC1

hgy59 commented:
@smaarn please wait with the update of homeassistant package as I am currently creating v0.118.5-11 with fixes for some integrations (and with python 3.7.7), see #4345.

smaarn commented:
> @smaarn please wait with the update of homeassistant package as I am currently creating v0.118.5-11 with fixes for some integrations (and with python 3.7.7), see #4345.

Don't worry I still need to try all the updated packages before pretending that this is ready to be merged.

smaarn commented:
Updated binaries can now be found here: https://github.com/smaarn/spksrc/releases/tag/python-3.8.6-RC2

Squashed icu commits and reorganized commits.

hgy59 commented:
> Squashed icu commits and reorganized commits.

@smaarn please avoid unnecessairy force-pushes as it always breaks local clones from fast-forwarding.

hgy59 commented:
@ymartin59 @smaarn and all py3 contirbutors here
As update of python 3.x to 3.(x+1) breaks all python3 packages installed with 3.x, we should consider to build different packages for 3.7 and 3.8.
If we start this versioning here we will leave python3 untouched and introduce python38 with it's own revision numbering (i.e. this would lead to python38-*-3.8.6-1.spk.

Benefits:
- do not break packages that do not support python 3.8 yet
- avoid the need of updating all py3 packages now
- do not break installed python3 packages at installation of python 3.8
- allow to create and publish python39 without retiring python38

_I am not a python guru but all seems to depend on the virtual environments that are linked to installed Python-Major.Minor._

smaarn commented:
> @ymartin59 @smaarn and all py3 contirbutors here
> 
> As update of python 3.x to 3.(x+1) breaks all python3 packages installed with 3.x, we should consider to build different packages for 3.7 and 3.8.
> 
> If we start this versioning here we will leave python3 untouched and introduce python38 with it's own revision numbering (i.e. this would lead to python38-*-3.8.6-1.spk.
> 
> 
> 
> Benefits:
> 
> - do not break packages that do not support python 3.8 yet
> 
> - avoid the need of updating all py3 packages now
> 
> - do not break installed python3 packages at installation of python 3.8
> 
> - allow to create and publish python39 without retiring python38
> 
> 
> 
> _I am not a python guru but all seems to depend on the virtual environments that are linked to installed Python-Major.Minor._
> 
> 

I would tend to agree with you here. Unfortunately I'm not an expert either here.
Another point to take into consideration is that other cross packages could need the same kind of approach (packages which no longer support python2 for example)

hgy59 commented on cross/PyNaCl/Makefile:
```diff
@@ -1,12 +1,11 @@
 PKG_NAME = PyNaCl
-PKG_VERS = 1.3.0
+PKG_VERS = 1.4.0
```
but homeassistant requires this in version 1.3.0
If we really need v1.4.0 for pythion 3.8 we must add cross/PyNaCl_1.3 and let homeassistant depend on it.

smaarn commented:
@hgy59 with your previous point in my mind I would propose to:
1. Create separate PRs for fixes included in this PR (znc, homeassistant and so on)
2. Create in another PR a new python38 package and let package owners perform their migration when they see fit (I already have a pending package which I could use as guinea pig)

What do you think?

iandarbey commented:
Any ideas when this will be merged? Some home assistant integrations rely on 3.8 or higher python (ikea in particular).

smaarn commented:
Superseeded by #4413
